src/components/shapes/draw.js

```diff
@@ -579,6 +599,224 @@ function setupDragElement(gd, shapePath, shapeOptions, index, shapeLayer, editHe
+            }
+        }
+    } else {
+        // Otherwise, we use the x and y bounds defined in the shape options
+        // and convert them to pixel coordinates
+        // Setup conversion functions
+        var xa = Axes.getFromId(gd, options.xref);
+        var xRefType = Axes.getRefType(options.xref);
+        var ya = Axes.getFromId(gd, options.yref);
+        var yRefType = Axes.getRefType(options.yref);
+        var x2p = helpers.getDataToPixel(gd, xa, false, xRefType);
+        var y2p = helpers.getDataToPixel(gd, ya, true, yRefType);
+        shapex0 = x2p(options.x0);
+        shapex1 = x2p(options.x1);
+        shapey0 = y2p(options.y0);
+        shapey1 = y2p(options.y1);
+    }
+
+    // Handle `auto` angle
+    var textangle = options.label.textangle;
+    if(textangle === 'auto') {
+        if(options.type === 'line') {
+            // Auto angle for line is same angle as line
+            textangle = calcTextAngle(shapex0, shapey0, shapex1, shapey1);
+        } else {
+            // Auto angle for all other shapes is 0
+            // The default angle for other shapes is 0 anyway, so there's no reason to
+            // explicitly set the angle to `auto`, this is just covering an edge case
+            textangle = 0;
+        }
+    }
+
+    // Do a fake render so we can get the text bounding box height
+    var _labelText = labelGroup.append('text')
+        .attr(labelTextAttrs)
+        .classed('shape-label-text', true)
+        .text(text);
+    _labelText.call(function(s) {
+        s.call(Drawing.font, font).attr({});
+        svgTextUtils.convertToTspans(s, gd);
+        return s;
+    });
+    var textBB = Drawing.bBox(_labelText.node());
+    _labelText.remove();
+
+    // Calculate correct (x,y) for text
+    // We also determine true xanchor since xanchor depends on position when set to 'auto'
+    var textPos = calcTextPosition(shapex0, shapey0, shapex1, shapey1, options, textangle, textBB);
+    var textx = textPos.textx;
+    var texty = textPos.texty;
+    var xanchor = textPos.xanchor;
+
+    function textLayout(s) {
+        s.call(Drawing.font, font)
+            .attr({
+                'text-anchor': {
+                    left: 'start',
+                    center: 'middle',
+                    right: 'end'
+                }[xanchor],
+                'y': texty,
+                'x': textx,
+                'transform': 'rotate(' + textangle + ',' + textx + ',' + texty + ')'
+            });
+        svgTextUtils.convertToTspans(s, gd);
+        return s;
+    }
+
+    labelText.call(textLayout);
+}
+
+function calcTextAngle(shapex0, shapey0, shapex1, shapey1) {
+    var dy, dx;
+    dx = Math.abs(shapex1 - shapex0);
+    if(shapex1 >= shapex0) {
+        dy = shapey0 - shapey1;
+    } else {
+        dy = shapey1 - shapey0;
+    }
+    return -180 / Math.PI * Math.atan2(dy, dx);
+}
+
+function calcTextPosition(shapex0, shapey0, shapex1, shapey1, shapeOptions, actualTextAngle, textBB) {
```

Comment:
Using `actualTextAngle` makes me wonder what's the difference between the two?
Looks like the only difference is the unit.
```suggestion
function calcTextPosition(shapex0, shapey0, shapex1, shapey1, shapeOptions, textAngle, textBB) {
```

Reply:
@archmoj The difference is that `actualTextAngle` is always a number -- it's the angle after `auto` has been converted to a real angle [here](https://github.com/plotly/plotly.js/blob/40933d044daa4e0ccf6a4bbdfaa5a7d4a639850b/src/components/shapes/draw.js#L671).

I agree it's a little confusing though, open to suggestions for making it clearer.